ddtrace/internal/module.py

```diff
@@ -0,0 +1,427 @@
+            except (ImportError, AttributeError):
+                loader = importlib.find_loader(fullname, path)
+            if loader:
+                return _ImportHookChainedLoader(loader, self.after_import)
+
+        finally:
+            self._finding.remove(fullname)
+
+        return None
+
+    def load_module(self, fullname):
+        # type: (str) -> ModuleType
+        # Python 2 only
+        module = sys.modules[fullname]
+        self.after_import(module)
+
+        return module
+
+    @classmethod
+    def register_origin_hook(cls, origin, hook, arg):
+        # type: (str, HookType, Any) -> None
+        """Register a hook to be called when the module with the given origin is
+        imported.
+
+        The hook will be called with the module object and the given argument.
+        """
+        assert isinstance(sys.modules, cls), "%r is installed" % cls
+
+        # DEV: Under the hypothesis that this is only ever called by the probe
+        # poller thread, there are no further actions to take. Should this ever
+        # change, then thread-safety might become a concern.
+        path = _resolve(origin)
+        if path is None:
+            raise ValueError("Cannot resolve module origin %s" % origin)
+
+        log.debug("Registering hook '%r' on path '%s'", hook, path)
+        sys.modules._hook_map[path].append((hook, arg))
+        try:
+            module = sys.modules._origin_map[path]
+        except KeyError:
+            # The module is not loaded yet. Nothing more we can do.
+            return
+
+        # The module was already imported so we invoke the hook straight-away
+        log.debug("Calling hook '%r' on already imported module '%s'", hook, module.__name__)
+        hook(module, arg)
+
+    @classmethod
+    def unregister_origin_hook(cls, origin, arg):
+        # type: (str, Any) -> None
+        """Unregister the hook registered with the given module origin and
+        argument.
+        """
+        assert isinstance(sys.modules, cls), "%r is installed" % cls
+
+        path = _resolve(origin)
+        if path is None:
+            raise ValueError("Module origin %s cannot be resolved", origin)
+
+        if path not in sys.modules._hook_map:
+            raise ValueError("No hooks registered for origin %s" % origin)
+
+        for i, (_, a) in enumerate(sys.modules._hook_map[path]):
+            if isinstance(a, list) and arg in a:
+                # DEV: This comes from the knowledge that sometimes the argument
+                # of the hook can be a list.
+                a.remove(arg)
+                if not a:
+                    sys.modules._hook_map[path].pop(i)
+                return
+            elif a is arg:
+                sys.modules._hook_map[path].pop(i)
+                return
+        raise ValueError("No hook registered for origin %s with argument %r" % (origin, arg))
+
+    @classmethod
+    def register_module_hook(cls, module, hook, arg):
```

Comment:
Maybe these could be module-level functions? (Again with the thinking that having public functions on something built-in feels weird)

Reply:
No longer relevant since we've decided that it's ok to introduce public methods on this class